tools/clang/lib/CodeGen/CGHLSLMS.cpp

```diff
@@ -7035,28 +7043,60 @@ void CGMSHLSLRuntime::EmitHLSLOutParamConversionInit(
         } else {
           continue;
         }
+      } else if (isAggregateType) {
+        // aggregate in-only - emit RValue, unless LValueToRValue cast
```

Comment:
Why? It seems that you would precisely want to emit an RValue if you have an LValueToRValue cast in order to respect copy semantics. Won't that make it so that `Call(LValueToRValue(DeclRef))` for an in argument will be able to modify DeclRef?

Reply:
Because this code and other code dependent on it is extremely sensitive to exactly what you do in each scenario.  If you attempt to emit as RValue here, various LValue cases will not emit properly, or assert, because we didn't add the appropriate HLSL-specific code to that path.  Instead of trying to fix all that, which is non-trivial, and add more regression risk to this change, I'm narrowing the code path to what we had before in the cases that don't impact the bugs I'm fixing here.  In a future larger change, we can rewrite this stuff and make it all better.

Remember, if we don't `continue` here, we will still copy the argument value in below (unless it's an object, but that's another conversation... ugh).

Reply:
Ok, I suspected as much. I agree with keeping it this way in the interest of minimizing churn and regression risk. I suspect we'll end up rewriting all of this code soon enough anyways.